### PR TITLE
Intermediate script: as soon as first MBs are in, you can run an intermediate script

### PR DESCRIPTION
**Pull request by sanderjo**
Handy to avoid downloads you don't want after all.

For example: as soon as the first MBs of a MKV are in, you can check for certain characteristics (like audio and subtitles) in the MKV.  Method: with this PR, SABnzbd will call your user defined intermediate_script, with the "working" directory as parameter, and your intermediate_script should the 0, 1, or -1 (or other SABnzbd known prio, like Pause, Forced, etc)

This works for

- traditional rar-set, with DirectUnpack on (aka True)
- no rar-set, with final file(s) in plain sight. Often: an xpost

Example intermediate script (hacked together!): 
https://github.com/sanderjo/anything_everything/blob/main/SABnzbd_post_pre_intermediate_scripts/mkv_checker.py

Result: a download automatically upvoted, one down voted, very early in the download phase.

image

### CODE REVIEW

**Comment by sanderjo** [comment]
Feedback wanted:

* calling & parsing the external intermediate_script is in two locations. Put it into one function?
* If assembler.py would know the extraction_path of DirectUnpack (in the nzo?), handling in one location would suffice.

**Comment by sanderjo** [comment]
Decision process logging for a post without rar-set, resp with rar-set & directunpacker on:

Something we want, because the external intermediate_script determines so:
```
2024-06-23 13:17:09,302::DEBUG::[assembler:150] SJ decision 1
2024-06-23 13:17:09,302::DEBUG::[assembler:151] SJ prio was 0
2024-06-23 13:17:09,302::DEBUG::[assembler:153] SJ prio is 1
```

Something we don't want (with tradtional rar-set post):
```
2024-06-23 13:17:57,136::DEBUG::[directunpacker:492] SJ decision -1
2024-06-23 13:17:57,136::DEBUG::[directunpacker:493] SJ prio was 0
2024-06-23 13:17:57,136::DEBUG::[directunpacker:495] SJ prio is -1
```

**Review by Safihre — COMMENTED** [review]
It should probably be the assembler that triggers the logic. But it should of course only be in 1 place.

File: sabnzbd/assembler.py [hunk 1]
```diff
@@ -80,6 +81,79 @@ def run(self):
...
+
+                        # TODO: first check if cfg.intermediate_script() and not nzo.intermediate_has_run
+
+                        if nzo.bytes_downloaded > 500_000_000:
+                            logging.debug("SJ 500 MB downloaded!")
+                            # if DirectUnpack is True, and has done some work, and
+                            # if we only knew the <complete_dir> and actual _UNPACK_ sub directory here,
+                            # ... we could handle it here
+
+                        if is_a_rarset:
```
[the remark this answers is not in the document]
Comment (sanderjo) [reply sabnzbd/assembler.py:114]: So ... if there is a first *rar: force unrar it to a temp dir, and then run the intermediate script on it?

**Pull request closed without merging**